tempodb/backend/versioned.go

```diff
@@ -0,0 +1,61 @@
+package backend
+
+import (
+	"context"
+	"io"
+
+	"github.com/pkg/errors"
+)
+
+type UpdateFn func(current io.ReadCloser) ([]byte, error)
+
+type Version string
+
+const (
+	// VersionNew is a placeholder version for a new file
+	VersionNew Version = "0"
+)
+
+var ErrVersionDoesNotMatch = errors.New("version does not match")
+
+// VersionedReaderWriter is a collection of methods to read and write data from tempodb backends with
+// versioning enabled.
+type VersionedReaderWriter interface {
+	RawReader
+
+	// WriteVersioned data to an object, if the version does not match the request will fail with
+	// ErrVersionDoesNotMatch. If the operation will create a new file, pass VersionNew.
+	WriteVersioned(ctx context.Context, name string, keypath KeyPath, data io.Reader, version Version) (Version, error)
+
+	// DeleteVersioned an object, if the version does not match the request will fail with
+	// ErrVersionDoesNotMatch.
+	DeleteVersioned(ctx context.Context, name string, keypath KeyPath, version Version) error
+
+	// ReadVersioned data from an object and returns the current version.
+	ReadVersioned(ctx context.Context, name string, keypath KeyPath) (io.ReadCloser, Version, error)
+}
+
+type fakeVersionedReaderWriter struct {
+	RawReader
+	RawWriter
+}
+
+var _ VersionedReaderWriter = (*fakeVersionedReaderWriter)(nil)
+
+func NewFakeVersionedReaderWriter(r RawReader, w RawWriter) VersionedReaderWriter {
```

Comment:
Consider returning the struct instead of the interface.

Reply:
That's a good tip that I almost always forget about. I have the tendency to hide abstractions from my Java days, but I agree returning the struct gives more flexibility to the caller since they can create their own interface around it.

For this specific example: `fakeVersionedReaderWriter` is private and returning it gives me this warning in GoLand. Not sure if making it public or just returning the interface would be best here? 🤔😅 

![Screenshot 2023-07-26 at 15 40 06](https://github.com/grafana/tempo/assets/7748404/0aa8bce6-f03f-41f4-b803-0c6bbfe8842e)

Reply:
I think making it public is okay.